### DISCUSSION TITLE
Restricting Workspace Access by User

### DISCUSSION THREAD

**Discussion by joshkimmel16** [eclipse-theia/theia#8391] [General]
Hello,

I'm still learning the ins and outs of theia so I apologize in advance if this question is obvious or there is a paradigm/feature that I'm missing. I wasn't able to find an existing question that is similar but if there is, I'd appreciate a pointer to it!
 
I'm looking to host theia on a remote server such that many different users can access it. For each user, I'm hoping to:
1) Allow that user to create new workspaces and toggle between the workspaces that he/she has created (as I understand it, the files associated with these workspaces would live on the remote server).
2) NOT allow that user to see/access the workspaces of other users.

What would be the best way to use/extend theia to accomplish this? Happy to provide more details if necessary.

**Comment by akosyakov** [comment]
Theia does not support multi tenancy. The primary use case one backend one user. This user can share the backend with other users for collaborative work though. For multi tenancy you will need to look into containerised solutions or somehow manage it with system rights.

A user can create a new workspace using File menu, i.e. Save as workspace. Previously opened workspaces are available via `Open recent workspaces` command. Workspace if either a folder (single-root workspace) or a manifest file for multi-root workspace. A user can have a folder containing all such folders and files and use File menu to switch between them.

**Reply by akosyakov** [reply]
You can also search the community forum: https://community.theia-ide.org for multi tenancy and restricting access. There were good discussions in the past.

**Reply by joshkimmel16** [reply]
Thanks for the pointer! As a quick follow up, at a high level, how much of the guts of Theia would I have to re-write if I were to fork/shim it and try to support multi-tenancy more formally? Just trying to get a sense of the effort that you think would be required to make such a change. I understand that the question is rather vague and entirely depends on more detailed requirements of what constitutes multi-tenancy, but any feedback you have would be greatly appreciated!

**Reply by akosyakov** [reply]
As I said Theia can handle multiple users. The issue with shared system resources. You will need to resolve on the system level. Not sure what you want to rewrite.
